### PR TITLE
VFX Graph gizmo disappear when trying to select it in the scene

### PR DESCRIPTION
**Pull request by julienamsellem**
### Purpose of this PR
https://fogbugz.unity3d.com/f/cases/1398351/

This is a regression introduced by this [PR](https://github.com/Unity-Technologies/Graphics/pull/6664).
As soon as the VFX Graph window loses focus the block is unselected and then the gizmo is removed from the scene.

Unity_GeYumhkSBG

---

### Testing status

Tested the original issues about selection from this [PR](https://github.com/Unity-Technologies/Graphics/pull/6664)
Tested operators and blocks gizmos.

---
### Comments to reviewers
Changelog is not updated because this issue has not been released to end user (or at worst in an alpha version)

### CODE REVIEW

**Review by MarieGuff — APPROVED** [review]
No T.A. work needed on this PR (bugfix only)

File: com.unity.visualeffectgraph/Editor/GraphView/Views/VFXPaste.cs [hunk 1]
```diff
@@ -80,9 +80,15 @@ static bool CanPasteNode(Node node, string openedAssetPath)
...
+            var content = data?.ToString();
+            if (string.IsNullOrEmpty(content))
+            {
+                return false;
+            }
+            
             try
             {
-                var serializableGraph = JsonUtility.FromJson<SerializableGraph>(data.ToString());
+                var serializableGraph = JsonUtility.FromJson<SerializableGraph>(content);
```
Comment (PaulDemeulenaere) [inline com.unity.visualeffectgraph/Editor/GraphView/Views/VFXPaste.cs:91]: The formatting is failing on this file 🤔
You can run it manually using:
```
perl $HOME\unity-meta\Tools\Format\format.pl --nobackups --hgroot $(pwd) .
```

Comment (PaulDemeulenaere) [reply com.unity.visualeffectgraph/Editor/GraphView/Views/VFXPaste.cs:91]: resolve at https://github.com/Unity-Technologies/Graphics/pull/6864/commits/a012ffe1d5208836a58fd819354c688f3c49ef00

**Review by PaulDemeulenaere — APPROVED** [review]
Reviewed the code (not tested)
- ```if (newValue != null)``` <= Only a safe null check 🟢 
- Noticed the change between OnLostFocus => UpdateGlobalSelection 🟢 (but can't know without testing)
- Unrelated safety check in `CanPaste` 🟢 : You can fill a fogbugz if you think it's approriate.

**Comment by julienamsellem** [comment]
> * Unrelated safety check in `CanPaste` 🟢 : You can fill a fogbugz if you think it's approriate.
I just had an exception, nobody reported it, so I thought I could quickly fix it without add too much "administrative" work 🙄

**Comment by VitaVFX** [comment]
Great and quick catch Julien! And also a great lesson for me 🙇🏻 

Tested:

❓**Connecting to Subgraphs Operators, then gizmos are not shown at all (I guess it is a known limitation according to provided warning message). Leaving for visibility**
image
image
image

✔️Gizmos of the property 
Unity_hpowvoqUo2

✔️Copy pasting/duplicating/deleting blocks

✔️ Blackboard properties gizmos are editable in Scene View

✔️**Gizmos between multi windows**
Unity_ShGnvrTHKu

✔️**Single-multi gizmos selection**
Unity_DqpuY01Yoq

✔️**Bounds recording, auto, manual settings**
Unity_9WFyylQD0t

✔️**Show Bounds in Visual Effect Panel in the Scene View**
Unity_IA9Q5kn37X

**Review by VitaVFX — APPROVED** [review]

**Comment by julienamsellem** [comment]
Thanks for the big test coverage 👍

**Comment by VitaVFX** [comment]
Cross-Posting for visibility. Several concerns based on suggestion.

1. Selecting empty space, brings us to the currently active VFX folder
2. Selecting empty space also focuses VFX asset not only shows it in the Inspector

highlighting the asset when clicking the empty space is very confusing. Is there a way only display asset's inspector without selecting the asset itself?

Unity_U8OYEEytON
Unity_Qwpk5x6mL1
n

**Comment by julienamsellem** [comment]
Thanks for catching this annoying UX behavior. Sadly there's no way to select the asset in the inspector and avoid the project browser from navigating to the asset. So now when clicking in the void in the VFX Graph the inspector is emptied if a vfx object was previously selected.

**Comment by VitaVFX** [comment]
<h3> Concerns: </h3>

1. We are no longer brought to the selection of VFX in Project Browser (yay!), but the fix also introduces one off inconsistency. So within this feature we also get a few concerns: 

- One off inconsistency among other windows
- Possible user confusion (might get some bug reports in the future regarding this). 

We probably want to weight what is impact of this implementation.

**[Observe Project's Browser Lock in gif]**
Unity_tS2cH5oUgL

2. Observation/Question. Is clicking into the 'void' supposed to shows Inspector only after when deselecting node/context/block? What about if I select something in Project Browser and then selecting 'void' in VFX Graph (without selecting any block/node/context)?
Unity_PySAjAbwfA

<h3>Passed Tests: </h3>

1. Selecting nodes/blocks/contexts display their information; selecting edges ports, shows info of previous selection
2. When opening external windows and clicking in the void of external asset does not bring us to VFX Asset window (that's goodie!)
3. Locking Inspector works fine. When Clicking blocks/nodes/contexts and locking Inspector their information is kept and vice versa
4. Right-clicking in the void shows information of previous selection
5. Quickly checked if gizmos can still be edited in Scene View by checking Arc Cone's and Bounds gizmos
6. Checked selection multi windows. If selection from contexts/node/block to the void brings any other heavier unwanted behavior.

**Comment by julienamsellem** [comment]
Thanks @VitaVFX for this new round of testing.
About the concerns you raised, here are my answers:
1/ The "lock" icon blink is the price to pay if we want to keep that behavior (unless someone knows a different way)
2/ If you have something selected in the project browser that is then also selected in the inspector, then clicking in the void of the VFX Graph won't discard this selection. That behavior is done on purpose and can be changed if it makes no sense.

**Comment by VitaVFX** [comment]
1/ Yes, I hear you and it is understandable. The problem is whether we want/can to pay this price. It would good to reach out to UX people and ask about this one off thingie. (I wish we could do this in a invisible way)

2/ Gotcha!

**Comment by julienamsellem** [comment]
FYI I found this bug report which was about the question we have about selecting the asset and the project browser changing directory https://fogbugz.unity3d.com/f/cases/1364098/

**Review by VitaVFX — APPROVED** [review]
[#vfx-ux conversation](https://unity.slack.com/archives/C02KRFTMN9W/p1644253336109809)

A sum up of the discussion for other reviewers:

1. This change is temporary, feature will adopt full intentional behavior within transition to GTF.
2. This feature has closest implementation to what GTF offers.
3. UX concerns:
             - Asset selection is associated to clicking into the void space. This change affects other areas.
             - Lock highlight most likely will be unnoticed, but if it will users most likely think that it is a bug.

Test results are above.

**Review by PaulDemeulenaere — APPROVED** [review]
Beside the (maybe) unexpected [using](https://github.com/Unity-Technologies/Graphics/pull/6864#discussion_r809255713) looks good to me.
Be sure it doesn't introduce a new failure in this [job](https://unity-ci.cds.internal.unity3d.com/job/11725955)

**Comment by MarieGuff** [comment]
**Found issues:** tested on 22.2.0a6.935

- **ABSOLUTE BLOCKER**: VFX's scene view control panel not available, probably replaced by gizo panel? 

VFX_Bug_GizmoPanel

- Gizmo error for non relevant items isn't necessary (ex: output selection, lifetime note selection...) they are also all displayed for a multiple selection

image

- Velocity gizmo not aligned with the arrow (probably unrelated to PR) 

(Expected: the orientation gizmo follows the arrow's direction)
VFX_Bug_VelocityGizmo

**Review by MarieGuff — CHANGES_REQUESTED** [review]
Bugs found, to be checked.

File: com.unity.visualeffectgraph/Editor/GraphView/Views/VFXView.cs [hunk 2]
```diff
@@ -2,13 +2,13 @@
 using System.Collections;
 using System.Collections.ObjectModel;
 using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
```
[the remark this answers is not in the document]
Comment (julienamsellem) [reply com.unity.visualeffectgraph/Editor/GraphView/Views/VFXView.cs]: Not required, thank!

**Comment by VitaVFX** [comment]
Verified raised issues and ran additional Test Rail pass to ensure UI functionality works as expected. [Test Doc](https://docs.google.com/document/d/1cY2U5A5k_qPKjQYp_Vi5cqHYQtpcf_fpAjRnUmmk_UI/edit#)

**Comment by VitaVFX** [comment]
@MarieGuff Would you confirm if fixes are suitable? Is there anything else concerning on TA's behalf?

**Comment by MarieGuff** [comment]
I still have the nodes that don't have a gizmo showing in the gizmo overlay with an error, can we change that @julienamsellem ? (ex: lifetime node don't have a gizmo so should not throw a "gizmo value is indeterminate" alert, same for color, size, orient... etc)

**Comment by julienamsellem** [comment]
> I still have the nodes that don't have a gizmo showing in the gizmo overlay with an error, can we change that @julienamsellem ? (ex: lifetime node don't have a gizmo so should not throw a "gizmo value is indeterminate" alert, same for color, size, orient... etc)

IIRC this should be addressed by a PR from @PaulDemeulenaere. Let's see with him when he is back :)

**Comment by MarieGuff** [comment]
Also It seems like I have to double click in the "void" to be able to create a new node with space bar, only one click should be enough

**Comment by julienamsellem** [comment]
> Also It seems like I have to double click in the "void" to be able to create a new node with space bar, only one click should be enough

This one must be addressed in this PR. I'm on it.

**Review by MarieGuff — APPROVED** [review]
Issues fixed thanks Julien!
TBD once Paul is back: Gizmo errors for non relevant nodes

**Comment by julienamsellem** [comment]
> Issues fixed thanks Julien! TBD once Paul is back: Gizmo errors for non relevant nodes

To be safe I also added auto-tests. They should catch any regression on focus when clearing the selection (click in the void for exemple)

**Comment by PaulDemeulenaere** [comment]
> 

Indeed, in this PR : https://ono.unity3d.com/unity/unity/pull-request/140065/_/graphics/vfx/feature/add-space-passthrough
I replaced with another message "gizmoWarningNotAvailable" which indicates "There isn't any gizmo available."

File: com.unity.visualeffectgraph/Editor/Types/VFXCircleGizmos.cs [hunk 3]
```diff
@@ -48,7 +48,7 @@ public static void DrawCircle(VFXGizmo gizmo, TCircle circle, IProperty<Vector3>
                         EditorGUI.BeginChangeCheck();
                         var dir = s_RadiusDirections[i];
                         var sliderPos = dir * radius;
-                        var result = Handles.Slider(s_RadiusDirectionsName[i], sliderPos, dir, handleSize * HandleUtility.GetHandleSize(sliderPos), CustomCubeHandleCap, 0.0f);
+                        var result = Handles.Slider(gizmo.GetCombinedHashCode(s_RadiusDirectionsName[i]), sliderPos, dir, handleSize * HandleUtility.GetHandleSize(sliderPos), CustomCubeHandleCap, 0.0f);
```
[the remark this answers is not in the document]
Comment (PaulDemeulenaere) [reply com.unity.visualeffectgraph/Editor/Types/VFXCircleGizmos.cs:51]: You should check usage of `GUIUtility.hotControl` to verify if we aren't compareing `s_RadiusDirectionsName` & co somewhere else.

File: com.unity.visualeffectgraph/Editor/Gizmo/VFXGizmo.cs [hunk 4]
```diff
@@ -358,10 +359,12 @@ public static void DefaultAngleHandleDrawFunction(int controlID, Vector3 positio
             rotation = Quaternion.LookRotation(tangent, normal);
             Matrix4x4 matrix = Matrix4x4.TRS(worldPosition, rotation, (Vector3.one + Vector3.forward * arcHandleSizeMultiplier));
             using (new Handles.DrawingScope(matrix))
-                Handles.CylinderHandleCap(controlID, Vector3.zero, Quaternion.identity, size, eventType);
+                Handles.CylinderHandleCap(GetCombinedHashCode(controlID), Vector3.zero, Quaternion.identity, size, eventType);
         }
 
         public virtual bool needsComponent { get { return false; } }
+
+        public int GetCombinedHashCode(int hashCode) => (currentHashCode + hashCode) % int.MaxValue;
```
[the remark this answers is not in the document]
Comment (julienamsellem) [reply com.unity.visualeffectgraph/Editor/Gizmo/VFXGizmo.cs]: Good point, I didn't know this method 👍

File: com.unity.visualeffectgraph/Editor/Types/VFXCircleGizmos.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (julienamsellem) [reply com.unity.visualeffectgraph/Editor/Types/VFXCircleGizmos.cs:51]: You were right, there was a small side effect on `hotControl` detection. It's fixed, but I think we should improve a bit later what we're doing with `hotControl`.

**Pull request closed without merging**